internal/exp/metrics/metrics.go

```diff
@@ -0,0 +1,123 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics"
+
+import (
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+)
+
+// Merge will merge the metrics data in mdB into mdA, then return mdA.
+// mdB will not be modified. The function will attempt to merge the data in mdB into
+// existing ResourceMetrics / ScopeMetrics / Metrics in mdA if possible. If they don't
```

Comment:
what are the benefits of merging it this way, instead of just adding mdB to a new resource metric on mdA? why would I want to use this, instead of just doing that? what are the perf implications? And why would this operation be here and not part of `pdata` itself?

Reply:
> what are the benefits of merging it this way, instead of just adding mdB to a new resource metric on mdA? why would I want to use this, instead of just doing that? 

To reduce data duplication during serialization. In the loadbalancingexporter, we split the incoming metric data by metric. Say our batch is 1 resource metric, 1 scope metric, and 4000 metrics. If we don't merge the resulting data, we end up with 4000 copies of the resource metric attributes and 4000 copies of the scope metric attributes. Which all has to be serialized to OTLP wire format.

By merging, we can reduce that back down to just 1 resource metric and 1 scope metric per group of metrics.

> What are the perf implications?

More CPU during processing time, to save CPU and bandwidth for the next step in the chain. It *should* be possible to measure it.

> And why would this operation be here and not part of pdata itself?
I assume you mean `pdatautil`? No good reason. If you feel it would be better there, then I can move it there instead.

Reply:
@open-telemetry/collector-approvers , I would need a second opinion here: would it be worth having this function as part of the pdata utilities, instead of having it here?

Reply:
Bump

Reply:
@jpkrohling in pdata you can merge without the need of a groupBy resource since that is faster. I don't think we can have a complicated logic like this directly in pdata, but if anything we can have a separate helper/util package with logic like this.

Reply:
Alright, in that case, I think it's OK to have it here in contrib as this PR proposes, and we can "promote" it to core if needed.